### DISCUSSION TITLE
[Exploit] Unlimited power ;)

### DISCUSSION THREAD

**Discussion by hexagonrecursion** [colobot/colobot#1665] [Workshop]
Usually a CBot script is limited to only running a small-ish number of instructions per frame, but there is a subset of CBot that can do math, has a conditional operator and a repetition operator that is capable of doing a **practically unlimited amount of work** in a single frame. Here is an example of incrementing a counter 7331 (1337 backwards) times:
```C++
// Tested in colobot version 0.2.1-alpha

public class Counter
{
	static int i = 0;
	static int limit = 0;
}

extern void Hax()
{
	Counter i();
	i.i = 0;
	i.limit = 7331;
	float t = abstime();
	try {
		Foo10000();
	}
	if(abstime() == t)
	{
		message("Successfully incremented counter " + i.i + " times in one frame");
	}
}

void Foo10000()
{
	Foo1000();
	Foo1000();
	Foo1000();
	Foo1000();
	Foo1000();
	
	Foo1000();
	Foo1000();
	Foo1000();
	Foo1000();
	Foo1000();
}

void Foo1000()
{
	Foo100();
	Foo100();
	Foo100();
	Foo100();
	Foo100();
	
	Foo100();
	Foo100();
	Foo100();
	Foo100();
	Foo100();
}

void Foo100()
{
	Foo10();
	Foo10();
	Foo10();
	Foo10();
	Foo10();
	
	Foo10();
	Foo10();
	Foo10();
	Foo10();
	Foo10();
}

void Foo10()
{
	Foo1();
	Foo1();
	Foo1();
	Foo1();
	Foo1();
	
	Foo1();
	Foo1();
	Foo1();
	Foo1();
	Foo1();
}

void Foo1()
{
	Counter i();
	// Division operator + try operator == a conditional operator
	1 / (i.limit - i.i);
	i.i++;
}
```

I leave figuring out what else this exploit can do as an exercise for the reader ;)

**Comment by melex750** [comment]
I think it is preferable to report something like this as steps to reproduce a **bug**, possibly in an existing issue.  In terms of development it is a bug.  In terms of playing the game it's considered cheating.  We should probably not advertise an "exploit" implying that it is a good thing or might be useful.
